### PR TITLE
Make Core init lazy, revise FileAccessProperty interface, fix file locking

### LINKED ISSUE #959: Use of H5P_SET_FILE_LOCKING env var is unreliable
We should use https://portal.hdfgroup.org/display/HDF5/H5P_SET_FILE_LOCKING rather than setting the environment variable `HDF5_USE_FILE_LOCKING`. Environment variable communication with the HDF5 library can be inconsistent, especially on Windows.

https://github.com/JuliaIO/HDF5.jl/blob/3a805bfddedfdac742161bc115323fcaf0d4e77a/src/HDF5.jl#L1710-L1712

Overall this indicates the need for configurable default access property lists.

### PR DESCRIPTION
**Pull request by mkitti**
* Fix #959 by using `h5p_set_file_locking(false, true)` by default via `HDF5.init!(fapl::FileAccessProperties)`
* Add `h5p_get_file_locking` helper, integrate with property interface
* Make `HDF5.Drivers.Core()` registration in `DRIVERS` lazy
* Implement `do` syntax for `Properties`
* Moves some initialization to `API.__init__()` which always executes first.
* Exposes a `const Ref` `libhdf5handle`
* Implements `API._has_symbol`

Revises #967

### CODE REVIEW

File: src/api/api.jl [hunk 1]
```diff
@@ -16,4 +16,24 @@ include("error.jl")
...
 include("helpers.jl")
 
+function __init__()
+    # HDF5.API.__init__() is run before HDF5.__init__()
+
+    # From deps.jl
+    check_deps()
+
+    # Ensure this is reinitialized on using
+    libhdf5handle[] = Libdl.dlopen(libhdf5)
```
Comment (musm) [inline src/api/api.jl:26]: This is needed to ensure the handle doesn't change between precompilation ?

File: src/api/helpers.jl [hunk 2]
```diff
@@ -285,6 +285,30 @@ function h5f_get_free_sections(file_id, type, sect_info::AbstractVector{H5F_sect
...
+   return _has_h5p_set_file_locking
+   #=
+   h5_version = h5_get_libversion()
+   if (h5_version >= v"1.10" && h5_version < v"1.10.7") ||
+      (h5_version >= v"1.12" && h5_version < v"1.12.1") ||
+      (h5_version < v"1.10")
+      return false
+   else
+      return true
+   end
```
Comment (musm) [inline src/api/helpers.jl:308]: The commented out lines are an alternative way to check this, but you opted for the `_has_symbol` approach instead?

**Comment by musm** [comment]
Are there any significant  runtime differences with the lazy approach?

**Comment by mkitti** [comment]
> Are there any significant  runtime differences with the lazy approach? 

HDF5.jl should load faster.

I suppose if anyone were to create a loop over setting the driver, it might be a little slower. The workaround would be to call the low level API directly.

File: src/api/helpers.jl (hunk 2 above)
Comment (mkitti) [reply src/api/helpers.jl:308]: Yes. We could do it either by version or looking up the symbol. Symbols lookup seems more direct, but the version based mechanism may still be informative.

File: src/api/api.jl (hunk 1 above)
Comment (musm) [reply src/api/api.jl:26]: > In the future, we should just grab the handle from HDF5_jll.

How would that that actually reinitialize the library? It seems like you'd have to call `dlopen` here regardless

**Comment by mkitti** [comment]
HDF_jll already uses `dlopen` for us via JLLWrappers.jl:
https://github.com/JuliaPackaging/JLLWrappers.jl/blob/bade7d4b7af3e1c0c590bfd383a3df5ed10852f8/src/products/library_generators.jl#L54

```
julia> HDF5.API.HDF5_jll.libhdf5_handle
Ptr{Nothing} @0x00000000009df580
```

**Comment by musm** [comment]
Ah nvm,  I misunderstood what you meant...

**Pull request merged**